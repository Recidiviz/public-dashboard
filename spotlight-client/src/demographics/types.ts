// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2021 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

export type TotalIdentifier = "ALL";
export const TOTAL_KEY: TotalIdentifier = "ALL";

type NoFilterIdentifier = "nofilter";
export const NOFILTER_KEY: NoFilterIdentifier = "nofilter";

export const UNKNOWN_KEY = "EXTERNAL_UNKNOWN";
type UnknownIdentifier = typeof UNKNOWN_KEY;

export const RaceOrEthnicityValueList = [
  "AMERICAN_INDIAN_ALASKAN_NATIVE",
  "BLACK",
  "HISPANIC",
  "WHITE",
  "ASIAN",
  "NATIVE_HAWAIIAN_PACIFIC_ISLANDER",
  "OTHER",
] as const;
export type RaceOrEthnicityValue = typeof RaceOrEthnicityValueList[number];
export type RaceIdentifier =
  | TotalIdentifier
  | UnknownIdentifier
  | RaceOrEthnicityValue;

export const GenderValueList = ["MALE", "FEMALE"] as const;
export type GenderValue = typeof GenderValueList[number];
export type GenderIdentifier =
  | TotalIdentifier
  | UnknownIdentifier
  | GenderValue;

export const AgeValueList = ["<25", "25-29", "30-34", "35-39", "40<"] as const;
export type AgeValue = typeof AgeValueList[number];
export type AgeIdentifier = TotalIdentifier | UnknownIdentifier | AgeValue;

export const DemographicViewList = [
  "total",
  "raceOrEthnicity",
  "gender",
  "ageBucket",
  "nofilter",
] as const;
export type DemographicView = typeof DemographicViewList[number];
export function isDemographicView(x: string): x is DemographicView {
  // because of how the array is typed, `includes` only accepts values
  // it already knows are in the array, which ... kind of defeats the purpose
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return DemographicViewList.includes(x as any);
}
