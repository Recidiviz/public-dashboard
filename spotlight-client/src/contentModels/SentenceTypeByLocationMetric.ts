// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2021 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { sum } from "d3-array";
import { computed, makeObservable } from "mobx";
import { SENTENCE_TYPE_LABELS } from "../constants";
import {
  getDemographicCategories,
  recordIsTotalByDimension,
} from "../demographics";
import {
  recordMatchesLocality,
  SentenceTypeByLocationRecord,
} from "../metricsApi";
import countUnknowns from "./countUnknowns";
import Metric, { BaseMetricConstructorOptions } from "./Metric";
import { UnknownCounts } from "./types";

type GraphNode = {
  id: string;
};

type GraphEdge = {
  source: string;
  target: string;
  value: number;
};

export default class SentenceTypeByLocationMetric extends Metric<
  SentenceTypeByLocationRecord
> {
  constructor(
    props: BaseMetricConstructorOptions<SentenceTypeByLocationRecord>
  ) {
    super(props);

    makeObservable(this, { dataGraph: computed, unknowns: computed });
  }

  get records(): SentenceTypeByLocationRecord[] | undefined {
    let recordsToReturn = this.allRecords;
    if (!recordsToReturn) return undefined;

    recordsToReturn = recordsToReturn.filter(
      recordMatchesLocality(this.localityId)
    );

    recordsToReturn = recordsToReturn.filter(
      recordIsTotalByDimension(this.demographicView)
    );

    return recordsToReturn;
  }

  get dataGraph():
    | { sources: GraphNode[]; targets: GraphNode[]; edges: GraphEdge[] }
    | undefined {
    const { demographicView, records } = this;
    if (!records || demographicView === "nofilter") return undefined;

    const sources = [
      SENTENCE_TYPE_LABELS.INCARCERATION,
      SENTENCE_TYPE_LABELS.PROBATION,
      SENTENCE_TYPE_LABELS.DUAL_SENTENCE,
    ];

    const categories = getDemographicCategories(demographicView);

    const edges = categories
      .map(({ identifier, label }) => {
        let record: SentenceTypeByLocationRecord | undefined;
        if (demographicView === "total") {
          // there's only one!
          [record] = records;
        } else {
          record = records.find((r) => r[demographicView] === identifier);
        }
        return [
          {
            source: sources[0],
            target: label,
            value: Number(record?.incarcerationCount) || 0,
          },
          {
            source: sources[1],
            target: label,
            value: Number(record?.probationCount) || 0,
          },
          {
            source: sources[2],
            target: label,
            value: Number(record?.dualSentenceCount) || 0,
          },
        ];
      })
      .reduce((flat, val) => flat.concat(val));

    return {
      sources: sources.map((name) => ({
        id: name,
      })),
      targets: categories.map(({ label }) => ({
        id: label,
      })),
      edges,
    };
  }

  get unknowns(): UnknownCounts | undefined {
    const { allRecords, localityId } = this;

    if (!allRecords) return undefined;

    return countUnknowns(
      allRecords.filter(recordMatchesLocality(localityId)),
      (groupedRecords: SentenceTypeByLocationRecord[]) =>
        sum(
          groupedRecords,
          (r) => r.dualSentenceCount + r.incarcerationCount + r.probationCount
        )
    );
  }
}
