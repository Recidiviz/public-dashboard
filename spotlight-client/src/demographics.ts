// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2021 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { DemographicFieldKey, DemographicView, TOTAL_KEY } from "./metricsApi";

export const DEMOGRAPHIC_UNKNOWN = "EXTERNAL_UNKNOWN";

const AGE_KEYS = {
  under25: "<25",
  "25_29": "25-29",
  "30_34": "30-34",
  "35_39": "35-39",
  over40: "40<",
};

const AGES = new Map([
  [AGE_KEYS.under25, "<25"],
  [AGE_KEYS["25_29"], "25-29"],
  [AGE_KEYS["30_34"], "30-34"],
  [AGE_KEYS["35_39"], "35-39"],
  [AGE_KEYS.over40, "40<"],
  [DEMOGRAPHIC_UNKNOWN, "Unknown"],
]);

const GENDERS = new Map([
  ["FEMALE", "Female"],
  ["MALE", "Male"],
  [DEMOGRAPHIC_UNKNOWN, "Unknown"],
]);

const RACES = {
  nativeAmerican: "AMERICAN_INDIAN_ALASKAN_NATIVE",
  black: "BLACK",
  hispanic: "HISPANIC",
  white: "WHITE",
  other: "OTHER",
};

// TODO: additional categories that weren't in ND?
const RACE_LABELS = new Map([
  [RACES.nativeAmerican, "Native American"],
  [RACES.black, "Black"],
  [RACES.hispanic, "Hispanic"],
  [RACES.white, "White"],
  [RACES.other, "Other"],
]);

export const DIMENSION_MAPPINGS = new Map<
  Exclude<DemographicView, "nofilter">,
  Map<string, string>
>([
  ["gender", GENDERS],
  ["age", AGES],
  ["race", RACE_LABELS],
  ["total", new Map([[TOTAL_KEY, "Total"]])],
]);

export const DIMENSION_DATA_KEYS: {
  [key in Extract<
    DemographicView,
    "race" | "gender" | "age" | "total"
  >]: DemographicFieldKey;
} = {
  race: "raceOrEthnicity",
  gender: "gender",
  age: "ageBucket",
  // TODO: this shouldn't ever happen but there is type weirdness
  total: "gender",
};
