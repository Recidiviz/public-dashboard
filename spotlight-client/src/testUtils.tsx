// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2020 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { LocationProvider } from "@reach/router";
import { render } from "@testing-library/react";
import React from "react";
import { autorun } from "mobx";
import waitForLocalhost from "wait-for-localhost";

// eslint-disable-next-line import/prefer-default-export
export function waitForTestServer(): Promise<void> {
  return waitForLocalhost({ path: "/health", port: 3002 });
}

/**
 * Convenience method to run an immediate, one-time reactive effect
 */
export function reactImmediately(effect: () => void): void {
  // this will call the effect function immediately,
  // and then immediately call the disposer to tear down the reaction
  autorun(effect)();
}

const LocationContextWrapper: React.FC = ({ children }) => {
  return <LocationProvider>{children}</LocationProvider>;
};

/**
 * Convenience method for rendering components that use @reach/router hooks
 * in a LocationContext to prevent rendering errors.
 */
export const renderWithRouter = (
  ui: React.ReactElement
): ReturnType<typeof render> => {
  return render(ui, { wrapper: LocationContextWrapper });
};
