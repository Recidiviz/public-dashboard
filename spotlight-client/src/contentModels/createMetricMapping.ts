// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2021 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { assertNever } from "assert-never";
import { MetricTypeIdList, TenantContent, TenantId } from "../contentApi/types";
import {
  parolePopulationCurrent,
  parolePopulationHistorical,
  paroleProgramParticipationCurrent,
  paroleRevocationReasons,
  paroleSuccessRateDemographics,
  paroleSuccessRateMonthly,
  prisonAdmissionReasons,
  prisonPopulationCurrent,
  prisonPopulationHistorical,
  prisonReleaseTypes,
  prisonStayLengths,
  probationPopulationCurrent,
  probationPopulationHistorical,
  probationProgramParticipationCurrent,
  probationRevocationReasons,
  probationSuccessRateDemographics,
  probationSuccessRateMonthly,
  recidivismRateAllFollowup,
  recidivismRateConventionalFollowup,
  sentencePopulationCurrent,
  sentenceTypesCurrent,
} from "../metricsApi";
import { MetricMapping } from "./types";
import DemographicsByCategoryMetric from "./DemographicsByCategoryMetric";
import HistoricalPopulationBreakdownMetric from "./HistoricalPopulationBreakdownMetric";
import PopulationBreakdownByLocationMetric from "./PopulationBreakdownByLocationMetric";
import ProgramParticipationCurrentMetric from "./ProgramParticipationCurrentMetric";
import RecidivismRateMetric from "./RecidivismRateMetric";
import SentenceTypeByLocationMetric from "./SentenceTypeByLocationMetric";
import SupervisionSuccessRateDemographicsMetric from "./SupervisionSuccessRateDemographicsMetric";
import SupervisionSuccessRateMonthlyMetric from "./SupervisionSuccessRateMonthlyMetric";
import { ERROR_MESSAGES } from "../constants";
import { NOFILTER_KEY, TOTAL_KEY } from "../demographics";
import { colors } from "../UiLibrary";

type MetricMappingFactoryOptions = {
  localityLabelMapping: TenantContent["localities"];
  metadataMapping: TenantContent["metrics"];
  tenantId: TenantId;
};
/**
 * Factory function for converting a mapping of content objects by metric ID
 * to a mapping of Metric instances by metric ID. Creating the entire mapping at once
 * ensures that each ID maps to the proper Metric type without requiring further
 * type guarding on the part of consumers.
 */
export default function createMetricMapping({
  localityLabelMapping,
  metadataMapping,
  tenantId,
}: MetricMappingFactoryOptions): MetricMapping {
  const metricMapping: MetricMapping = new Map();

  // to maintain type safety we iterate through all of the known metrics;
  // iterating through the metadata object's keys widens the type to `string`,
  // which prevents us from guaranteeing exhaustiveness at the type level
  MetricTypeIdList.forEach((metricType) => {
    // not all metrics are required; metadata object is the source of truth
    // for which metrics to include
    const metadata = metadataMapping[metricType];
    if (!metadata) {
      return;
    }

    // messages for errors we may hit at various points below
    const localityContentError = `${ERROR_MESSAGES.missingRequiredContent} (locality labels)`;
    const totalLabelError = `${ERROR_MESSAGES.missingRequiredContent} (total population label)`;

    // this big ol' switch statement ensures that the type ID string union is properly narrowed,
    // allowing for 1:1 correspondence between the ID and the typed Metric instance.
    switch (metricType) {
      case "SentencePopulationCurrent":
        if (!localityLabelMapping?.Sentencing)
          throw new Error(localityContentError);

        if ("totalLabel" in metadata)
          metricMapping.set(
            metricType,
            new PopulationBreakdownByLocationMetric({
              ...metadata,
              id: metricType,
              tenantId,
              defaultDemographicView: NOFILTER_KEY,
              defaultLocalityId: TOTAL_KEY,
              localityLabels: localityLabelMapping.Sentencing,
              dataTransformer: sentencePopulationCurrent,
              sourceFileName: "sentence_type_by_district_by_demographics",
            })
          );
        else throw new Error(totalLabelError);

        break;
      case "SentenceTypesCurrent":
        if (!localityLabelMapping?.Sentencing)
          throw new Error(localityContentError);

        metricMapping.set(
          metricType,
          new SentenceTypeByLocationMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: TOTAL_KEY,
            localityLabels: localityLabelMapping.Sentencing,
            dataTransformer: sentenceTypesCurrent,
            sourceFileName: "sentence_type_by_district_by_demographics",
          })
        );
        break;
      case "PrisonPopulationCurrent":
        if (!localityLabelMapping?.Prison)
          throw new Error(localityContentError);

        if ("totalLabel" in metadata)
          metricMapping.set(
            metricType,
            new PopulationBreakdownByLocationMetric({
              ...metadata,
              id: metricType,
              tenantId,
              defaultDemographicView: NOFILTER_KEY,
              defaultLocalityId: TOTAL_KEY,
              localityLabels: localityLabelMapping.Prison,
              dataTransformer: prisonPopulationCurrent,
              sourceFileName:
                "incarceration_population_by_facility_by_demographics",
            })
          );
        else throw new Error(totalLabelError);
        break;
      case "ProbationPopulationCurrent":
        if (!localityLabelMapping?.Probation)
          throw new Error(localityContentError);
        if ("totalLabel" in metadata)
          metricMapping.set(
            metricType,
            new PopulationBreakdownByLocationMetric({
              ...metadata,
              id: metricType,
              tenantId,
              defaultDemographicView: NOFILTER_KEY,
              defaultLocalityId: TOTAL_KEY,
              localityLabels: localityLabelMapping.Probation,
              dataTransformer: probationPopulationCurrent,
              sourceFileName:
                "supervision_population_by_district_by_demographics",
            })
          );
        else throw new Error(totalLabelError);
        break;
      case "ParolePopulationCurrent":
        if (!localityLabelMapping?.Parole)
          throw new Error(localityContentError);
        if ("totalLabel" in metadata)
          metricMapping.set(
            metricType,
            new PopulationBreakdownByLocationMetric({
              ...metadata,
              id: metricType,
              tenantId,
              defaultDemographicView: NOFILTER_KEY,
              defaultLocalityId: TOTAL_KEY,
              localityLabels: localityLabelMapping.Parole,
              dataTransformer: parolePopulationCurrent,
              sourceFileName:
                "supervision_population_by_district_by_demographics",
            })
          );
        else throw new Error(totalLabelError);
        break;
      case "PrisonPopulationHistorical":
        metricMapping.set(
          metricType,
          new HistoricalPopulationBreakdownMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: prisonPopulationHistorical,
            sourceFileName: "incarceration_population_by_month_by_demographics",
          })
        );
        break;
      case "ProbationPopulationHistorical":
        metricMapping.set(
          metricType,
          new HistoricalPopulationBreakdownMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: probationPopulationHistorical,
            sourceFileName: "supervision_population_by_month_by_demographics",
          })
        );
        break;
      case "ParolePopulationHistorical":
        metricMapping.set(
          metricType,
          new HistoricalPopulationBreakdownMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: parolePopulationHistorical,
            sourceFileName: "supervision_population_by_month_by_demographics",
          })
        );
        break;
      case "ProbationProgrammingCurrent":
        if (!localityLabelMapping?.Probation)
          throw new Error(localityContentError);

        metricMapping.set(
          metricType,
          new ProgramParticipationCurrentMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: undefined,
            defaultLocalityId: NOFILTER_KEY,
            localityLabels: localityLabelMapping.Probation,
            dataTransformer: probationProgramParticipationCurrent,
            sourceFileName: "active_program_participation_by_region",
          })
        );
        break;
      case "ParoleProgrammingCurrent":
        if (!localityLabelMapping?.Parole)
          throw new Error(localityContentError);

        metricMapping.set(
          metricType,
          new ProgramParticipationCurrentMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: undefined,
            defaultLocalityId: NOFILTER_KEY,
            localityLabels: localityLabelMapping.Parole,
            dataTransformer: paroleProgramParticipationCurrent,
            sourceFileName: "active_program_participation_by_region",
          })
        );
        break;
      case "ProbationSuccessHistorical":
        if (!localityLabelMapping?.Probation)
          throw new Error(localityContentError);

        metricMapping.set(
          metricType,
          new SupervisionSuccessRateMonthlyMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: undefined,
            defaultLocalityId: TOTAL_KEY,
            localityLabels: localityLabelMapping.Probation,
            dataTransformer: probationSuccessRateMonthly,
            sourceFileName: "supervision_success_by_month",
          })
        );
        break;
      case "ParoleSuccessHistorical":
        if (!localityLabelMapping?.Parole)
          throw new Error(localityContentError);

        metricMapping.set(
          metricType,
          new SupervisionSuccessRateMonthlyMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: undefined,
            defaultLocalityId: TOTAL_KEY,
            localityLabels: localityLabelMapping.Parole,
            dataTransformer: paroleSuccessRateMonthly,
            sourceFileName: "supervision_success_by_month",
          })
        );
        break;
      case "ProbationSuccessAggregate":
        if (!localityLabelMapping?.Probation)
          throw new Error(localityContentError);

        metricMapping.set(
          metricType,
          new SupervisionSuccessRateDemographicsMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: TOTAL_KEY,
            localityLabels: localityLabelMapping.Probation,
            dataTransformer: probationSuccessRateDemographics,
            sourceFileName: "supervision_success_by_period_by_demographics",
          })
        );
        break;
      case "ParoleSuccessAggregate":
        if (!localityLabelMapping?.Parole)
          throw new Error(localityContentError);

        metricMapping.set(
          metricType,
          new SupervisionSuccessRateDemographicsMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: TOTAL_KEY,
            localityLabels: localityLabelMapping.Parole,
            dataTransformer: paroleSuccessRateDemographics,
            sourceFileName: "supervision_success_by_period_by_demographics",
          })
        );
        break;
      case "ProbationRevocationsAggregate":
        metricMapping.set(
          metricType,
          new DemographicsByCategoryMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: probationRevocationReasons,
            sourceFileName:
              "supervision_revocations_by_period_by_type_by_demographics",
          })
        );
        break;
      case "ParoleRevocationsAggregate":
        metricMapping.set(
          metricType,
          new DemographicsByCategoryMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: paroleRevocationReasons,
            sourceFileName:
              "supervision_revocations_by_period_by_type_by_demographics",
          })
        );
        break;
      case "PrisonAdmissionReasonsCurrent":
        metricMapping.set(
          metricType,
          new DemographicsByCategoryMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: prisonAdmissionReasons,
            sourceFileName: "incarceration_population_by_admission_reason",
          })
        );
        break;
      case "PrisonReleaseTypeAggregate":
        metricMapping.set(
          metricType,
          new DemographicsByCategoryMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: prisonReleaseTypes,
            sourceFileName: "incarceration_releases_by_type_by_period",
          })
        );
        break;
      case "PrisonRecidivismRateHistorical":
        metricMapping.set(
          metricType,
          new RecidivismRateMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: recidivismRateAllFollowup,
            sourceFileName: "recidivism_rates_by_cohort_by_year",
          })
        );
        break;
      case "PrisonRecidivismRateSingleFollowupHistorical":
        metricMapping.set(
          metricType,
          new RecidivismRateMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            followUpYears: 3,
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: recidivismRateConventionalFollowup,
            sourceFileName: "recidivism_rates_by_cohort_by_year",
          })
        );
        break;
      case "PrisonStayLengthAggregate":
        metricMapping.set(
          metricType,
          new DemographicsByCategoryMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: prisonStayLengths,
            sourceFileName: "incarceration_lengths_by_demographics",
            color: colors.dataVizNamed.get("teal"),
          })
        );
        break;
      default:
        assertNever(metricType);
    }
  });

  return metricMapping;
}
