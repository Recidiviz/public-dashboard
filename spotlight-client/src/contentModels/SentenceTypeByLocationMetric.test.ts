// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2021 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { runInAction, when } from "mobx";
import { DemographicView } from "../demographics";
import { reactImmediately } from "../testUtils";
import createMetricMapping from "./createMetricMapping";
import SentenceTypeByLocationMetric from "./SentenceTypeByLocationMetric";
import contentFixture from "./__fixtures__/tenant_content_exhaustive";

const testTenantId = "US_ND";
const testMetricId = "SentenceTypesCurrent";
const testMetadataMapping = {
  [testMetricId]: contentFixture.metrics[testMetricId],
};

function getTestMetric() {
  return createMetricMapping({
    localityLabelMapping: contentFixture.localities,
    metadataMapping: testMetadataMapping,
    tenantId: testTenantId,
  }).get(testMetricId) as SentenceTypeByLocationMetric;
}

async function getPopulatedMetric() {
  const metric = getTestMetric();

  metric.hydrate();

  await when(() => Boolean(metric.records));

  return metric;
}

test("total data", async () => {
  const metric = await getPopulatedMetric();

  reactImmediately(() => {
    expect(metric.dataGraph).toMatchSnapshot();
  });

  expect.hasAssertions();
});

test.each([["raceOrEthnicity"], ["gender"], ["ageBucket"]] as [
  Exclude<DemographicView, "nofilter">
][])("%s data", async (demographicView) => {
  const metric = await getPopulatedMetric();

  runInAction(() => {
    metric.demographicView = demographicView;
  });

  reactImmediately(() => {
    expect(metric.dataGraph).toMatchSnapshot();
  });

  expect.hasAssertions();
});

test("locality filter", async () => {
  const metric = await getPopulatedMetric();

  reactImmediately(() =>
    expect(metric.records?.every((record) => record.locality === "ALL")).toBe(
      true
    )
  );

  const facilityId = contentFixture.localities.Sentencing.entries[1].id;

  runInAction(() => {
    metric.localityId = facilityId;
  });

  reactImmediately(() => {
    expect(metric.records?.length).toBeGreaterThan(0);
    expect(
      metric.records?.every((record) => record.locality === facilityId)
    ).toBe(true);
  });

  expect.hasAssertions();
});
