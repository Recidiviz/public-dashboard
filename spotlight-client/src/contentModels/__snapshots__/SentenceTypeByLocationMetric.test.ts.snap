// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ageBucket data 1`] = `
Object {
  "edges": Array [
    Object {
      "source": "Incarceration",
      "target": "<25",
      "value": 1147,
    },
    Object {
      "source": "Probation",
      "target": "<25",
      "value": 96,
    },
    Object {
      "source": "Both",
      "target": "<25",
      "value": 267,
    },
    Object {
      "source": "Incarceration",
      "target": "25-29",
      "value": 1262,
    },
    Object {
      "source": "Probation",
      "target": "25-29",
      "value": 1248,
    },
    Object {
      "source": "Both",
      "target": "25-29",
      "value": 538,
    },
    Object {
      "source": "Incarceration",
      "target": "30-39",
      "value": 3295,
    },
    Object {
      "source": "Probation",
      "target": "30-39",
      "value": 604,
    },
    Object {
      "source": "Both",
      "target": "30-39",
      "value": 836,
    },
    Object {
      "source": "Incarceration",
      "target": "40-49",
      "value": 272,
    },
    Object {
      "source": "Probation",
      "target": "40-49",
      "value": 795,
    },
    Object {
      "source": "Both",
      "target": "40-49",
      "value": 369,
    },
    Object {
      "source": "Incarceration",
      "target": "50-59",
      "value": 3,
    },
    Object {
      "source": "Probation",
      "target": "50-59",
      "value": 55,
    },
    Object {
      "source": "Both",
      "target": "50-59",
      "value": 9,
    },
    Object {
      "source": "Incarceration",
      "target": "60-69",
      "value": 19,
    },
    Object {
      "source": "Probation",
      "target": "60-69",
      "value": 563,
    },
    Object {
      "source": "Both",
      "target": "60-69",
      "value": 32,
    },
    Object {
      "source": "Incarceration",
      "target": "70+",
      "value": 43,
    },
    Object {
      "source": "Probation",
      "target": "70+",
      "value": 188,
    },
    Object {
      "source": "Both",
      "target": "70+",
      "value": 6,
    },
  ],
  "sources": Array [
    Object {
      "id": "Incarceration",
    },
    Object {
      "id": "Probation",
    },
    Object {
      "id": "Both",
    },
  ],
  "targets": Array [
    Object {
      "id": "<25",
    },
    Object {
      "id": "25-29",
    },
    Object {
      "id": "30-39",
    },
    Object {
      "id": "40-49",
    },
    Object {
      "id": "50-59",
    },
    Object {
      "id": "60-69",
    },
    Object {
      "id": "70+",
    },
  ],
}
`;

exports[`gender data 1`] = `
Object {
  "edges": Array [
    Object {
      "source": "Incarceration",
      "target": "Male",
      "value": 6034,
    },
    Object {
      "source": "Probation",
      "target": "Male",
      "value": 3112,
    },
    Object {
      "source": "Both",
      "target": "Male",
      "value": 1961,
    },
    Object {
      "source": "Incarceration",
      "target": "Female",
      "value": 159,
    },
    Object {
      "source": "Probation",
      "target": "Female",
      "value": 285,
    },
    Object {
      "source": "Both",
      "target": "Female",
      "value": 96,
    },
  ],
  "sources": Array [
    Object {
      "id": "Incarceration",
    },
    Object {
      "id": "Probation",
    },
    Object {
      "id": "Both",
    },
  ],
  "targets": Array [
    Object {
      "id": "Male",
    },
    Object {
      "id": "Female",
    },
  ],
}
`;

exports[`raceOrEthnicity data 1`] = `
Object {
  "edges": Array [
    Object {
      "source": "Incarceration",
      "target": "Native American",
      "value": 1050,
    },
    Object {
      "source": "Probation",
      "target": "Native American",
      "value": 0,
    },
    Object {
      "source": "Both",
      "target": "Native American",
      "value": 225,
    },
    Object {
      "source": "Incarceration",
      "target": "Black",
      "value": 1613,
    },
    Object {
      "source": "Probation",
      "target": "Black",
      "value": 115,
    },
    Object {
      "source": "Both",
      "target": "Black",
      "value": 371,
    },
    Object {
      "source": "Incarceration",
      "target": "Hispanic",
      "value": 422,
    },
    Object {
      "source": "Probation",
      "target": "Hispanic",
      "value": 1581,
    },
    Object {
      "source": "Both",
      "target": "Hispanic",
      "value": 430,
    },
    Object {
      "source": "Incarceration",
      "target": "White",
      "value": 1946,
    },
    Object {
      "source": "Probation",
      "target": "White",
      "value": 377,
    },
    Object {
      "source": "Both",
      "target": "White",
      "value": 498,
    },
    Object {
      "source": "Incarceration",
      "target": "Other",
      "value": 1061,
    },
    Object {
      "source": "Probation",
      "target": "Other",
      "value": 1425,
    },
    Object {
      "source": "Both",
      "target": "Other",
      "value": 533,
    },
  ],
  "sources": Array [
    Object {
      "id": "Incarceration",
    },
    Object {
      "id": "Probation",
    },
    Object {
      "id": "Both",
    },
  ],
  "targets": Array [
    Object {
      "id": "Native American",
    },
    Object {
      "id": "Black",
    },
    Object {
      "id": "Hispanic",
    },
    Object {
      "id": "White",
    },
    Object {
      "id": "Other",
    },
  ],
}
`;

exports[`total data 1`] = `
Object {
  "edges": Array [
    Object {
      "source": "Incarceration",
      "target": "Total",
      "value": 6193,
    },
    Object {
      "source": "Probation",
      "target": "Total",
      "value": 3399,
    },
    Object {
      "source": "Both",
      "target": "Total",
      "value": 2056,
    },
  ],
  "sources": Array [
    Object {
      "id": "Incarceration",
    },
    Object {
      "id": "Probation",
    },
    Object {
      "id": "Both",
    },
  ],
  "targets": Array [
    Object {
      "id": "Total",
    },
  ],
}
`;
